### PR TITLE
[Inductor] Build Shape Padding in Inductor

### PR DESCRIPTION
**Pull request by jiawenliu64**
Summary: Build shape padding for matmul/bmm/addmm in Inductor

Differential Revision: D41071282

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Review by bertmaher — APPROVED** [review]

File: torch/_inductor/config.py [hunk 1]
```diff
@@ -67,6 +67,9 @@
 # How to import torchdynamo, either torchdynamo or torch.dynamo
 dynamo_import = inductor_import.replace("inductor", "dynamo")
 
+# Pad input tensors of matmul/bmm/addmm to leverage Tensor Cores in NVIDIA GPUs
+shape_padding = os.environ.get("TORCHINDUCTOR_SHAPE_PADDING", "0") == "1"
```
Comment (eellison) [inline torch/_inductor/config.py:71]: what's stopping us from turning this on by default ?

Comment (jiawenliu64) [reply torch/_inductor/config.py:71]: Hi @eellison, we tested shape padding optimization in a few internal and OSS models, and we will make it by default when the optimization is
- tested in most internal and OSS benchmarking models;
- supported other dtype (e.g., FP16 for inference) with different heuristics

File: torch/_inductor/decomposition.py [hunk 2]
```diff
@@ -144,8 +167,138 @@ def addmm(input, mat1, mat2, *, beta=1, alpha=1):
...
+                return torch.ops.aten.addmm(input, mat1, mat2, beta=beta, alpha=alpha)[
+                    :, :-(n_padded_length)
+                ]
+        elif k_padded_length != 0:
+            return torch.ops.aten.addmm(input, mat1, mat2, beta=beta, alpha=alpha)
+
+    return NotImplemented  # go directly to lowering
+
+
+def bench(f, name=None, iters=100, warmup=5, display=False):
```
Comment (jgong5) [inline torch/_inductor/decomposition.py:200]: Try to reuse the bench code in utils.py?
https://github.com/pytorch/pytorch/blob/master/torch/_inductor/utils.py#L106

File: torch/_inductor/decomposition.py [hunk 3]
```diff
@@ -144,8 +167,138 @@ def addmm(input, mat1, mat2, *, beta=1, alpha=1):
...
+
+        k_padded_length = get_padded_length(mat1.shape[2])
+        n_padded_length = get_padded_length(mat2.shape[2])
+
+        if k_padded_length != 0:
+            mat1 = pad_dim(mat1, k_padded_length, 2)
+            mat2 = pad_dim(mat2, k_padded_length, 1)
+        if n_padded_length != 0:
+            mat2 = pad_dim(mat2, n_padded_length, 2)
+            return torch.ops.aten.bmm(mat1, mat2)[:, :, :-n_padded_length].contiguous()
```
Comment (Chillee) [inline torch/_inductor/decomposition.py:297]: I added this contiguous call sinceI ran into needing it for correctness, but I think this might actually necessary in all cases?

cc: @ngimel 

It does sadly seem to impact perf a bit in my testing though.

**Review by Chillee — APPROVED** [review]
LGTM overall

File: torch/_inductor/decomposition.py (hunk 2 above)
Comment (jiawenliu64) [reply torch/_inductor/decomposition.py]: In `utils.py`, we add a new bench which is more accurate and reliable
